test/unit/test_quilbase.py

```diff
@@ -195,3 +196,73 @@ def test_classical_reg(self, measurement: Measurement, classical_reg: MemoryRefe
         assert measurement.classical_reg == classical_reg
         measurement.classical_reg = MemoryReference("new_mem_ref")
         assert measurement.classical_reg == MemoryReference("new_mem_ref")
+
+
+@pytest.mark.parametrize(
+    ("frame", "direction", "initial_frequency", "hardware_object", "sample_rate", "center_frequency"),
+    [
+        (Frame([Qubit(0)], "frame"), None, None, None, None, None),
+        (Frame([Qubit(1)], "frame"), "direction", 1.39, "hardware_object", 44.1, 440.0),
+    ],
+    ids=("Frame-Only", "With-Optionals"),
+)
+class TestDefFrame:
+    @pytest.fixture
+    def def_frame(
+        self,
+        frame: Frame,
+        direction: Optional[str],
+        initial_frequency: Optional[float],
+        hardware_object: Optional[str],
+        sample_rate: Optional[float],
+        center_frequency: Optional[float],
+    ):
+        optional_args = [
+            arg
+            for arg in [direction, initial_frequency, hardware_object, sample_rate, center_frequency]
+            if arg is not None
+        ]
+        return DefFrame(frame, *optional_args)
+
+    def test_out(self, def_frame: DefFrame, snapshot: SnapshotAssertion):
+        # The ordering of attributes isn't stable and can be printed in different orders across calls.
+        # We assert that the first line is definitely the `DEFFRAME` line, and that the following
+        # attributes are the same, regardless of their order.
+        quil_lines = def_frame.out().splitlines()
+        assert quil_lines[0] == snapshot
+        assert set(quil_lines[1:]) == snapshot
```

Comment:
This was an interesting one. `dict`s in Python maintain ordering, but `HashMap`s in Rust don't. Thus, once you pass a `dict` over the boundary to a `HashMap`, the order isn't guaranteed once it comes back.

Reply:
Also, besides the ordering of attribute values, there were no meaningful differences between the formatting of `pyQuil`'s `DefFrame` and `quil-rs`' `FrameDefinition`.

Reply:
If it were important to maintain ordering, we could use an `IndexMap`, but that's not called for in the spec so I think what you've done here is right